main/src/ca/uwaterloo/flix/language/phase/Redundancy.scala

```diff
@@ -267,72 +299,74 @@ object Redundancy extends Phase[TypedAst.Root, TypedAst.Root] {
 
     case Expression.Apply(exp1, exp2, _, _, _) =>
-      val us1 = visitExp(exp1, env0)
-      val us2 = visitExp(exp2, env0)
-      Used.empty ++ us1 ++ us2
+      val env1 = env0.addApply
+      val env2 = env0.resetApplies
+      val us1 = visitExp(exp1, env1)
+      val us2 = visitExp(exp2, env2)
+      us1 and us2
 
     case Expression.Unary(_, exp, _, _, _) =>
-      visitExp(exp, env0)
+      visitExp(exp, env0.resetApplies)
 
     case Expression.Binary(_, exp1, exp2, _, _, _) =>
-      val us1 = visitExp(exp1, env0)
-      val us2 = visitExp(exp2, env0)
-      us1 ++ us2
+      val us1 = visitExp(exp1, env0.resetApplies)
+      val us2 = visitExp(exp2, env0.resetApplies)
+      us1 and us2
 
     case Expression.Let(sym, exp1, exp2, _, _, _) =>
       // Extend the environment with the variable symbol.
       val env1 = env0 + sym
 
       // Visit the two expressions under the extended environment.
-      val innerUsed1 = visitExp(exp1, env1)
-      val innerUsed2 = visitExp(exp2, env1)
+      val innerUsed1 = visitExp(exp1, env1.resetApplies)
+      val innerUsed2 = visitExp(exp2, env1.resetApplies)
 
       // Check for shadowing.
       val shadowedVar = shadowing(sym, env0)
 
       // Check if the let-bound variable symbol is dead in exp2.
       if (deadVarSym(sym, innerUsed2))
-        (innerUsed1 ++ innerUsed2 ++ shadowedVar) - sym + UnusedVarSym(sym)
+        (innerUsed1 and innerUsed2 and shadowedVar) - sym + UnusedVarSym(sym)
       else
-        (innerUsed1 ++ innerUsed2 ++ shadowedVar) - sym
+        (innerUsed1 and innerUsed2 and shadowedVar) - sym
 
     case Expression.LetRec(sym, exp1, exp2, _, _, _) =>
       // Extend the environment with the variable symbol.
       val env1 = env0 + sym
 
       // Visit the two expressions under the extended environment.
-      val innerUsed1 = visitExp(exp1, env1)
-      val innerUsed2 = visitExp(exp2, env1)
+      val innerUsed1 = visitExp(exp1, env1.resetApplies)
+      val innerUsed2 = visitExp(exp2, env1.resetApplies)
 
       // Check for shadowing.
       val shadowedVar = shadowing(sym, env0)
 
       // Check if the let-bound variable symbol is dead in exp1 and exp2.
-      if (deadVarSym(sym, innerUsed1 ++ innerUsed2))
-        (innerUsed1 ++ innerUsed2 ++ shadowedVar) - sym + UnusedVarSym(sym)
+      if (deadVarSym(sym, innerUsed1 and innerUsed2))
+        (innerUsed1 and innerUsed2 and shadowedVar) - sym + UnusedVarSym(sym)
       else
-        (innerUsed1 ++ innerUsed2 ++ shadowedVar) - sym
+        (innerUsed1 and innerUsed2 and shadowedVar) - sym
 
     case Expression.IfThenElse(exp1, exp2, exp3, _, _, _) =>
-      val us1 = visitExp(exp1, env0)
-      val us2 = visitExp(exp2, env0)
-      val us3 = visitExp(exp3, env0)
-      us1 ++ us2 ++ us3
+      val us1 = visitExp(exp1, env0.resetApplies)
+      val us2 = visitExp(exp2, env0.resetApplies)
+      val us3 = visitExp(exp3, env0.resetApplies)
+      us1 and (us2 or us3)
```

Comment:
Hmm. I am trying to think.... Does this correctly handle:

```
def foo(x: Int, y: Int): Int = 
    (if (?) foo(1) else foo(1))(42)
```

and could it support that?

Reply:
It doesn't handle this correctly. I'm thinking we could instead count applies on the way up. Instead of storing a `boolean` isConditionallyRecursive we can store a count of recursive applications and then checking at the top if the count > arity. I'll include this in the next iteration if I can get it working.

Reply:
Hmmm. I don't know what to do. I think I want to simplest implementation that gets the most bang for the buck. I don't know if this case should be fixed or not.

Reply:
After some puzzling, it looks like this would require some significant changes. I've left it out of the latest revision, but I think it is doable.